src/main/java/build/buildfarm/server/ContentAddressableStorageService.java

```diff
@@ -129,13 +130,18 @@ public void getTree(
       return;
     }
     ImmutableList.Builder<Directory> directories = new ImmutableList.Builder<>();
-    String nextPageToken = instance.getTree(
-        request.getRootDigest(), pageSize, request.getPageToken(), directories);
 
-    responseObserver.onNext(GetTreeResponse.newBuilder()
-        .addAllDirectories(directories.build())
-        .setNextPageToken(nextPageToken)
-        .build());
+    String pageToken = request.getPageToken();
+    do {
+      String nextPageToken = instance.getTree(
+          request.getRootDigest(), pageSize, pageToken, directories);
```

Comment:
You want to clear directories between every onNext. I don't see https://github.com/uber/bazel-buildfarm/blob/working/src/main/java/build/buildfarm/instance/AbstractServerInstance.java#L183 doing that.

Also: while this is better than before, it could be made even more efficient by refactoring instance.getTree to return everything at once instead of repeatedly calling it here. The problem with calling it repeatedly is that you're repeatedly fetching the entire tree, and then slicing out just one portion of it. It's still O(n^2) overall, even though it's all in memory so you could ignore it, I guess.

Reply:
Missed the directories builder reset, thanks.

You should take another look at my implementation of things here - I'm not refetching the entire tree on each call to MemoryInstance::getTree - my page token contains a path hierarchy of digests and my restarts are logarithmic.